### PR TITLE
feat: session overflow model

### PR DESCRIPTION
**Pull request by romanetar**
ref https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskDueDate&view=vertical&task=3637557&stafilter=NotComplete&fileview=grid

Signed-off-by: romanetar <roman_ag@hotmail.com>

### CODE REVIEW

File: app/ModelSerializers/Summit/SummitEventOverflowStreamingSerializer.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+     * @param array $relations
+     * @param array $params
+     * @return array
+     */
+    public function serialize
+    (
+        $expand = null, array $fields = [], array $relations = [], array $params = []
+    ): array
+    {
+        $event = $this->object;
```
Comment (smarcet) [inline app/ModelSerializers/Summit/SummitEventOverflowStreamingSerializer.php:36]: use the fields array declaration , dont do this

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments

File: app/ModelSerializers/Summit/AdminSummitEventSerializer.php [hunk 2]
```diff
@@ -22,6 +22,9 @@ class AdminSummitEventSerializer extends SummitEventSerializer
 {
     protected static $array_mappings = [
         'Occupancy' => 'occupancy:json_string',
+        'OverflowStreamingUrl' => 'overflow_streaming_url:json_string',
```
Comment (smarcet) [inline app/ModelSerializers/Summit/AdminSummitEventSerializer.php:25]: please review implementation
this fields are missing from allowed fields and in_array test

File: app/ModelSerializers/Summit/ISummitEventSerializerTypes.php [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * limitations under the License.
+ **/
+
+/**
+ * Interface ISummitEventSerializerTypes
+ * @package ModelSerializers
+ */
+interface ISummitEventSerializerTypes
+{
+    const OverflowStream = 'OVERFLOW_STREAM';
```
Comment (smarcet) [inline app/ModelSerializers/Summit/ISummitEventSerializerTypes.php:21]: move this to IPresentationSerializerTypes

File: app/Models/Foundation/Summit/Events/SummitEvent.php [hunk 4]
```diff
@@ -1739,6 +1757,73 @@ public function promote2Presentation(PresentationType $type):void{
...
+            config("overflow.query_string_key","k"),
+            $this->overflow_stream_key);
+    }
+
+    /**
+     * @return string|null
+     */
+    public function getOverflowTokens(): ?string
+    {
+        return null;
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Events/SummitEvent.php:1827]: review implementation from here https://github.com/OpenStackweb/summit-api/blob/f3aa6ee51bd1fa5af5d4c36fe6d915d9b6d0d286/app/ModelSerializers/Summit/SummitEventSecureStreamSerializer.php#L109

File: app/Repositories/Summit/DoctrineSummitEventRepository.php [hunk 5]
```diff
@@ -842,6 +842,27 @@ public function getLastPresentationOrderBySummit(int $summit_id): int
         }
     }
 
+    /**
+     * feature/session_overflow
+     * @param string $enc_key
+     * @return SummitEvent|null
+     */
+    public function getByOverflowStreamKey(string $enc_key): ?SummitEvent
```
Comment (smarcet) [inline app/Repositories/Summit/DoctrineSummitEventRepository.php:850]: change enc_key to something more meaning full

File: app/Repositories/Summit/DoctrineSummitEventRepository.php [hunk 6]
```diff
@@ -842,6 +842,27 @@ public function getLastPresentationOrderBySummit(int $summit_id): int
...
+    {
+        try {
+            return $this->getEntityManager()->createQueryBuilder()
+                ->select("e")
+                ->from($this->getBaseEntity(), "e")
+                ->where('e.overflow_stream_key = :overflow_stream_key')
+                ->setParameter('overflow_stream_key', strtoupper($enc_key))
```
Comment (smarcet) [inline app/Repositories/Summit/DoctrineSummitEventRepository.php:857]: why to upper , use trim here

File: app/Models/Foundation/Summit/Factories/SummitEventFactory.php [hunk 7]
```diff
@@ -137,6 +137,14 @@ static public function populate(Summit $summit, SummitEvent $event, array $paylo
             $event->setSubmissionSource(trim($payload['submission_source']));
         }
 
+        if(isset($payload['overflow_streaming_url'])) {
+            $event->setOverflowStreamingUrl(trim($payload['overflow_streaming_url']));
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Factories/SummitEventFactory.php:141]: this is not correct please review the use

File: app/Models/Foundation/Summit/Events/SummitEvent.php [hunk 8]
```diff
@@ -1739,6 +1757,73 @@ public function promote2Presentation(PresentationType $type):void{
...
+     * @param string $overflow_stream_key
+     * @return void
+     */
+    public function setOverflow(string $overflow_streaming_url,
+                                bool $overflow_stream_is_secure,
+                                string $overflow_stream_key): void
+    {
+        $this->overflow_streaming_url = $overflow_streaming_url;
+        $this->overflow_stream_is_secure = $overflow_stream_is_secure;
+        $this->overflow_stream_key = $overflow_stream_key;
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Events/SummitEvent.php:1798]: key is not needed here
bc at factory level u don have it

File: app/Services/Model/Imp/SummitService.php [hunk 9]
```diff
@@ -330,6 +332,20 @@ public function __construct
...
+    /**
+     * @param SummitEvent $event
+     * @return void
+     * @throws RandomException
+     */
+    private function generateOverflowStreamKey(SummitEvent $event): string
+    {
+        do {
+            $salt = random_bytes(16);
+            $overflow_key = hash('sha256', $event->getId() . $salt . time());
```
Comment (smarcet) [inline app/Services/Model/Imp/SummitService.php:344]: this logic should be on SummitEvent

File: app/Services/Model/Imp/SummitService.php [hunk 10]
```diff
@@ -4160,4 +4176,63 @@ public function publishStreamUpdatesStartInXMinutes(int $minutes): void
...
+        return $this->tx_service->transaction(function () use ($summit, $event_id, $payload) {
+
+            $event = $this->event_repository->getByIdRefreshed($event_id);
+            if (is_null($event))
+                throw new ValidationException(sprintf("Event id %s does not exists,", $event_id));
+
+            $event->setOverflow(
+                $payload['overflow_streaming_url'],
+                boolval($payload['overflow_stream_is_secure']),
+                $this->generateOverflowStreamKey($event));
```
Comment (smarcet) [inline app/Services/Model/Imp/SummitService.php:4200]: this should be set on another call to be able to call setOverflow at factory level

File: app/Services/Model/Imp/SummitService.php [hunk 11]
```diff
@@ -4160,4 +4176,63 @@ public function publishStreamUpdatesStartInXMinutes(int $minutes): void
...
+            $params = [
+                'overflow_url' => $event->getOverflowUrl()
+            ];
+
+            ProcessScheduleEntityLifeCycleEvent::dispatch
+            (
+                "UPDATE",
+                $summit->getId(),
+                $event->getId(),
+                self::PresentationOverflowEntityType,
```
Comment (smarcet) [inline app/Services/Model/Imp/SummitService.php:4211]: this should be somewhere else

File: app/Services/Model/Imp/SummitService.php [hunk 12]
```diff
@@ -4160,4 +4176,63 @@ public function publishStreamUpdatesStartInXMinutes(int $minutes): void
...
+    /**
+     * @param Summit $summit
+     * @param int $event_id
+     * @return void
+     * @throws Exception
+     */
+    public function removeOverflowState(Summit $summit, int $event_id): SummitEvent
+    {
+        return $this->tx_service->transaction(function () use ($summit, $event_id) {
+
```
Comment (smarcet) [inline app/Services/Model/Imp/SummitService.php:4228]: add log debug info

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments

File: app/ModelSerializers/Summit/AdminSummitEventSerializer.php (hunk 2 above)
Comment (romanetar) [reply app/ModelSerializers/Summit/AdminSummitEventSerializer.php:25]: done

File: app/ModelSerializers/Summit/ISummitEventSerializerTypes.php (hunk 3 above)
Comment (romanetar) [reply app/ModelSerializers/Summit/ISummitEventSerializerTypes.php:21]: done

File: app/Models/Foundation/Summit/Events/SummitEvent.php (hunk 4 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Events/SummitEvent.php:1827]: done

File: app/Repositories/Summit/DoctrineSummitEventRepository.php (hunk 5 above)
Comment (romanetar) [reply app/Repositories/Summit/DoctrineSummitEventRepository.php:850]: done

File: app/Repositories/Summit/DoctrineSummitEventRepository.php (hunk 6 above)
Comment (romanetar) [reply app/Repositories/Summit/DoctrineSummitEventRepository.php:857]: done

File: app/Models/Foundation/Summit/Factories/SummitEventFactory.php (hunk 7 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Factories/SummitEventFactory.php:141]: done

File: app/Services/Model/Imp/SummitService.php (hunk 10 above)
Comment (romanetar) [reply app/Services/Model/Imp/SummitService.php:4200]: done

File: app/Services/Model/Imp/SummitService.php (hunk 11 above)
Comment (romanetar) [reply app/Services/Model/Imp/SummitService.php:4211]: done

File: app/Http/Controllers/Apis/Protected/Summit/Factories/SummitEventValidationRulesFactory.php [hunk 13]
```diff
@@ -69,6 +69,8 @@ public static function build(array $data, bool $update = false): array
                 'custom_order' => 'sometimes|integer',
                 'duration' => 'sometimes|integer|min:0',
                 'stream_is_secure' =>  'sometimes|boolean',
+                'overflow_stream_is_secure' => 'sometimes|boolean',
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/Factories/SummitEventValidationRulesFactory.php:72]: remove this

File: app/ModelSerializers/Summit/AdminSummitEventSerializer.php (hunk 2 above)
Comment (smarcet) [reply app/ModelSerializers/Summit/AdminSummitEventSerializer.php:25]: @romanetar is missing this part
https://github.com/OpenStackweb/summit-api/blob/e8ab6755771f18e95ab50c55baac1719cdfe6a90/app/ModelSerializers/Summit/AdminSummitEventSerializer.php#L72

File: app/ModelSerializers/Summit/SummitEventOverflowStreamingSerializer.php [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+class SummitEventOverflowStreamingSerializer extends AbstractSerializer
+{
+    protected static $array_mappings = [
+        'Id' => 'id:json_int',
+        'Title' => 'title:json_string',
+        'StartDate' => 'start_date:datetime_epoch',
+        'EndDate' => 'end_date:datetime_epoch',
+        'OverflowStreamingUrl' => 'overflow_streaming_url:json_string',
+        'OverflowStreamIsSecure' => 'overflow_stream_is_secure:json_boolean',
+        'OverflowTokens' => 'overflow_tokens:json_string',
```
Comment (smarcet) [inline app/ModelSerializers/Summit/SummitEventOverflowStreamingSerializer.php:30]: overflow_tokens is an array not a json_string

File: app/Models/Foundation/Summit/Events/SummitEvent.php [hunk 15]
```diff
@@ -1739,6 +1762,120 @@ public function promote2Presentation(PresentationType $type):void{
...
+    /**
+     * @return string
+     */
+    public function getOverflowUrl(): string
+    {
+        return sprintf("%s%s?%s=%s",
+            $this->summit->getMarketingSiteUrl(),
+            config("overflow.path", "/a/overflow-player"),
+            config("overflow.query_string_key","k"),
+            $this->overflow_stream_key);
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Events/SummitEvent.php:1830]: validate that key is not null 
throw a validatoin exception

File: app/Models/Foundation/Summit/Events/SummitEvent.php [hunk 16]
```diff
@@ -1739,6 +1762,120 @@ public function promote2Presentation(PresentationType $type):void{
...
+        $this->overflow_stream_key = null;
+        $this->occupancy = self::OccupancyEmpty;
+    }
+
+    /**
+     * @return string
+     */
+    public function getOverflowUrl(): string
+    {
+        return sprintf("%s%s?%s=%s",
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Events/SummitEvent.php:1826]: validate that occupancy is overflow

File: app/Models/Foundation/Summit/Events/SummitEvent.php [hunk 17]
```diff
@@ -1739,6 +1762,120 @@ public function promote2Presentation(PresentationType $type):void{
...
+    public function generateOverflowKey(): string
+    {
+         $salt = random_bytes(16);
+         return hash('sha256', $this->getId() . $salt . time());
+    }
+
+    /**
+     * @return array
+     */
+    public function getOverflowTokens(): array
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Events/SummitEvent.php:1846]: refactoring this to have an unique method
for overflow and for regular stream
please do check https://github.com/OpenStackweb/summit-api/blob/main/app/ModelSerializers/Summit/SummitEventSecureStreamSerializer.php
and unify the methods ( knowledge should be on one single place )

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php [hunk 18]
```diff
@@ -1549,4 +1548,108 @@ public function upgradeEvent($summit_id, $event_id, $type_id)
...
+    /**
+     * @param $summit_id
+     * @return \Illuminate\Http\JsonResponse|mixed
+     */
+    public function getOverflowStreaming($summit_id)
+    {
+        return $this->processRequest(function () use ($summit_id) {
+            $query_string_key = config("overflow.query_string_key", "k");
+            if (!Request::has($query_string_key)) return $this->error404();
+
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php:1561]: its 400 ( bad request)

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php [hunk 19]
```diff
@@ -1549,4 +1548,108 @@ public function upgradeEvent($summit_id, $event_id, $type_id)
...
+            Log::debug(sprintf("OAuth2SummitEventsApiController::updateOverflowInfo summit id %s event id %s", $summit_id, $event_id));
+
+            if (!Request::isJson()) return $this->error400();
+            $data = Request::json();
+
+            $summit = SummitFinderStrategyFactory::build($this->repository, $this->resource_server_context)->find($summit_id);
+            if (is_null($summit))
+                return $this->error404();
+
+            $event = $summit->getEvent($event_id);
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php:1602]: intval($event_id)

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php [hunk 20]
```diff
@@ -1549,4 +1548,108 @@ public function upgradeEvent($summit_id, $event_id, $type_id)
...
+            if (is_null($event))
+                return $this->error404();
+
+            $payload = $data->all();
+
+            $validation = Validator::make($payload, [
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php:1608]: move this to
app/Http/Controllers/Apis/Protected/Summit/Factories/SummitEventValidationRulesFactory.php
create a new static method to validate overflow info
and remove overflow fields from the regular methds

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php [hunk 21]
```diff
@@ -1549,4 +1548,108 @@ public function upgradeEvent($summit_id, $event_id, $type_id)
...
+            );
+        });
+    }
+
+    /**
+     * @param $summit_id
+     * @param $event_id
+     * @return mixed
+     */
+    public function updateOverflowInfo($summit_id, $event_id) {
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php:1590]: rename to setOverflow

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php [hunk 22]
```diff
@@ -1549,4 +1548,108 @@ public function upgradeEvent($summit_id, $event_id, $type_id)
...
+                ->serialize
+                (
+                    SerializerUtils::getExpand(),
+                    SerializerUtils::getFields(),
+                    SerializerUtils::getRelations(),
+                )
+            );
+        });
+    }
+    public function removeOverflowState($summit_id, $event_id)
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php:1631]: rename to clearOverflow

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php [hunk 23]
```diff
@@ -1549,4 +1548,108 @@ public function upgradeEvent($summit_id, $event_id, $type_id)
             );
         });
     }
+
+    /**
+     * @param $summit_id
+     * @return \Illuminate\Http\JsonResponse|mixed
+     */
+    public function getOverflowStreaming($summit_id)
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php:1556]: rename to getOverflowStreamingInfo

File: tests/BrowserKitTestCase.php [hunk 24]
```diff
@@ -55,21 +55,21 @@ protected function setUp():void
...
-        DB::setDefaultConnection("config");
-        DB::delete('DELETE FROM endpoint_api_scopes');
-        DB::delete('DELETE FROM endpoint_api_authz_groups');
-        DB::delete('DELETE FROM api_scopes');
-        DB::delete('DELETE FROM api_endpoints');
-        DB::delete('DELETE FROM apis');
-
-        $this->seed(ConfigSeeder::class);
-        $this->seed(MainDataSeeder::class);
+//        // see https://laravel.com/docs/9.x/mocking#mail-fake
```
Comment (smarcet) [inline tests/BrowserKitTestCase.php:58]: rolll back this

File: app/ModelSerializers/Summit/Presentation/AdminSummitEventCSVSerializer.php [hunk 25]
```diff
@@ -21,6 +21,9 @@ class AdminSummitEventCSVSerializer extends SummitEventSerializer
 {
     protected static $array_mappings = [
         'Occupancy' => 'occupancy:json_string',
+        'OverflowStreamingUrl' => 'overflow_streaming_url:json_string',
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/ModelSerializers/Summit/Presentation/AdminSummitEventCSVSerializer.php:24]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php (hunk 18 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php:1561]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php (hunk 21 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php:1590]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php (hunk 22 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php:1631]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php (hunk 23 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php:1556]: done

File: tests/BrowserKitTestCase.php (hunk 24 above)
Comment (romanetar) [reply tests/BrowserKitTestCase.php:58]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php (hunk 20 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEventsApiController.php:1608]: done

File: app/Models/Foundation/Summit/Events/SummitEvent.php (hunk 16 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Events/SummitEvent.php:1826]: done

File: app/Models/Foundation/Summit/Events/SummitEvent.php (hunk 15 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Events/SummitEvent.php:1830]: done

File: app/ModelSerializers/Summit/SummitEventOverflowStreamingSerializer.php (hunk 14 above)
Comment (romanetar) [reply app/ModelSerializers/Summit/SummitEventOverflowStreamingSerializer.php:30]: done

File: app/Models/Foundation/Summit/Events/SummitEvent.php [hunk 26]
```diff
@@ -1739,6 +1764,160 @@ public function promote2Presentation(PresentationType $type):void{
...
+    {
+         $salt = random_bytes(16);
+         return hash('sha256', $this->getId() . $salt . time());
+    }
+
+    /**
+     * @param bool $overflow
+     * @return array
+     */
+    public function getStreamingTokens(bool $overflow = true): array
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Events/SummitEvent.php:1856]: dont use flags
create 2 methods 
getStreamingTokens and getOverflowStreamingTokens

and encapsulate the core strategy on one method parametrizable method ( trait ? )

File: app/Models/Foundation/Summit/Events/SummitEvent.php [hunk 27]
```diff
@@ -1739,6 +1764,160 @@ public function promote2Presentation(PresentationType $type):void{
...
+    {
+        $tokens = [];
+        $summit = $this->summit;
+        $cache_key = $this->getSecureStreamCacheKey();
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Events/SummitEvent.php:1860]: the cache key is not the same 
u need to have one for regular stream and another one for overflow stream

File: app/Models/Foundation/Summit/Events/SummitEvent.php (hunk 26 above)
Comment (smarcet) [reply app/Models/Foundation/Summit/Events/SummitEvent.php:1856]: here , variant s, between regular and overflow stream
are
* streaming_url
* cache key 
the core logic is the same , so lets create private method that implements the core logic and have as param both variants and lets that private method be called by 
**getStreamingTokens** and **getOverflowStreamingTokens**

**Comment by smarcet** [comment]
LGTM

**Review by smarcet — APPROVED** [review]
LGTM

**Pull request merged**